### PR TITLE
Annotated Precision

### PR DESCRIPTION
**Pull request by fitnessflo**
At the moment the active `IntDomains` for an analysis can only be set for an entire program, e.g. with the use of `// PARAM: --enable ana.int.interval`. Activating more Domains for the entire analysis might lead to more precise results, but this can be expensive (more `evals`), when only a couple functions actually need this level of precision. Thus being able to decide for each function separately what `IntDomains` should be activated is desirable.

In this PR a feature to annotate each function with a different precision was implemented. By activating `// PARAM: --enable annotated.int.enabled` for an analysis a user can annotate a function with a custom C-attribute `__attribute__((goblint_precision(...)))`
and thus enable different `IntDomains`:

- `DefExc`
- `Interval`
- `Enums`
- `Congruence`

Since the active Domains change between function calls refinement has to be used to accommodate the change in precision. Thus `// PARAM: --set ana.int.refinement once/fixpoint` has to be activated as well **for the feature to work!**

# Precision attribute
The attribute `goblin_precision` can be used to fine-tune the active `IntDomains` for a function.
The following string arguments are supported:

1. `def_exc`/`no-def_exc`
2. `interval`/`no-interval`
3. `enums`/`no-enums`
4. `congruence`/`no-congruence`

This can be used in conjunction with the usual annotation of a file, e.g. `// PARAM: --enable ana.int.def_exc`.
This can be useful in cases where every function except one should activate a specific Domain. So instead of annotating every function, only one function has to be annotated with a `no-...` option.

This feature makes use of the `ContextUtil.should_keep` method. Since the underlying `Contextutil.has_attribute` method only checks the `Cil.attrparam list` for matching strings the risk of mismatching the actual attribute is now an issue. While the supported string arguments of `precision` and `goblint_context` differ new attributes might be added in the future. Hence `ContextUtil.should_keep` method has been modified to additionally take a variant type. Furthermore a helper function to get the corresponding strings has been added.
The current types are:

- `GobContext` -> `goblint_context`
- `GobPrecision` -> `goblint_precision`

So now `Contextutil.has_attribute` matches both the string for the attribute as well as the strings for the attrparams. For the correct use of attributes see `user-guide/annotating.md`.

## Changes

- Modefied `Contextutil`
- Added `util/precisionUtil.ml` to determine the precision, defined as a tuple of `(bool * bool * bool * bool)`,  from a `fundec` or a `node` in case of globals.
- `Enter/Combine` now `project` the values of the current `cpa` to the precision of the `Callee/Caller` if `annotation.int.enabled` is activated. If `annotation.int.privglobs` is activated they will further more update privatized globals to the highest precision.
- Added `ValueDomain.projection p v` which matches the compound type `v`:
  - `Int` will be projected by `IntDomTuple.projection p v`
  -  The rest will be recursively projected, e.g. a `struct` might have a field `int`
- Added `IntDomTupleImpl.projection ik p v`which refines `v` and then disables inactive Domains.
- In case `exp.earlyglobs` is enabled globals will be projected with the highest precision, to not lose information when used by functions with different precision.
- Privatized globals will now be projected to the highest precision, when both the flags `annotation.int.enabled` and `annotation.int.privglobs` are activated. (**Side Note**: We now encounter a case where a local variable will have the highest precision, when assigned the value of a privatized global. Though the precision will be corrected, when the local variable is called at a later program point)
- `create/create2` have been split into two functions each. the first type has the same signature but will construct the precision from a node or the config, depending if the flag `annotation.int.enabled` is set. The second type takes an additional argument `p` (precision), and constructs the tuple from the information.
- Added annotated versions of some `regtests` for evaluation

### CODE REVIEW

**Comment by sim642** [comment]
> Also, I think at least for the files with only whitespace changes, it might be a good idea to undo your changes, to keep the history somewhat more clean.

Even in files with changes, this is a major dealbreaker for `git blame` because all these whitespace changes are mixed into the same commits as actual changes, so it's not valid to add those commits even to `.git-blame-ignore-revs`.

File: src/analyses/base.ml [hunk 1]
```diff
@@ -1127,13 +1128,29 @@ struct
       if (!GU.earlyglobs || ThreadFlag.is_multi a) && is_global a x then begin
         if M.tracing then M.tracel "setosek" ~var:x.vname "update_one_addr: update a global var '%s' ...\n" x.vname;
         let new_value = update_offset (Priv.read_global a gs st x) in
+
+        (* Projection to highest Precision *)
+        let new_value =
+          if GobConfig.get_bool "annotation.int.enabled"
+          then VD.projection (PU.max_precision ()) new_value
+          else new_value
+        in
```
[the remark this answers is not in the document]
Comment (fitnessflo) [reply src/analyses/base.ml:1137]: So should I project the `new_value` inside `update_offset` instead?

File: src/cdomains/intDomain.mli [hunk 2]
```diff
@@ -312,6 +316,12 @@ module PtrDiffIkind : Ikind
 
 module IntDomWithDefaultIkind (I: Y) (Ik: Ikind) : Y with type t = I.t and type int_t = I.int_t
 
+module IntDomUtil : sig
+  val max_precision: unit -> (bool * bool * bool * bool)
```
[the remark this answers is not in the document]
Comment (jerhard) [reply src/cdomains/intDomain.mli]: @michael-schwarz It  is `max_precision` in the sense that all possible domains are activated. It does not depend on any configuration.

File: src/analyses/base.ml (hunk 1 above)
[the remark this answers is not in the document]
Comment (jerhard) [reply src/analyses/base.ml:1137]: One would project the `value` inside `update_offset`.

File: src/cdomains/intDomain.ml [hunk 3]
```diff
@@ -2563,12 +2581,31 @@ module IntDomTupleImpl = struct
...
   let create2 r x = (* use where values are introduced *)
-    let f n g = if get_bool ("ana.int."^n) then Some (g x) else None in
-    f "def_exc" @@ r.fi2 (module I1), f "interval" @@ r.fi2 (module I2), f "enums" @@ r.fi2 (module I3), f "congruence" @@ r.fi2 (module I4)
+    let p =
+      if GobConfig.get_bool "annotation.int.enabled" then
+        precision_from_node ()
+      else
+        let f n = get_bool ("ana.int."^n) in
+        (f "def_exc", f "interval", f "enums", f "congruence")
+    in
```
Comment (jerhard) [inline src/cdomains/intDomain.ml:2607]: Is this the same as for `create`? Could this be deduplicated?

**Review by jerhard — COMMENTED** [review]
Looks good to me now.

**Review by michael-schwarz — APPROVED** [review]

**Review by sim642 — CHANGES_REQUESTED** [review]
This creates a test directory `41-annotated-precision`, but on `master` there is already `41-stdlib`. The PR CI hasn't rerun, so the conflict wasn't detected.

**Comment by jerhard** [comment]
@fitnessflo Could you please rename the test folder from `41-annotated-precision` to `42-annotated-precision`?

**Review by sim642 — APPROVED** [review]

**Review by jerhard — APPROVED** [review]

**Pull request merged**